src/server/core.rs

```diff
@@ -1,96 +1,52 @@
 use std::rc::Rc;
 
-use crate::common::{IdCounter, Identifiable, KeyIdMap, Map, Set, WrappedRcRefCell};
+use crate::common::{IdCounter, Map, Set, WrappedRcRefCell};
 use crate::scheduler::{TaskAssignment, TaskId, WorkerId};
-use crate::server::client::{Client, ClientId};
-use crate::server::notifications::Notifications;
-use crate::server::protocol::key::{
-    dask_key_ref_to_str, dask_key_ref_to_string, DaskKey, DaskKeyRef,
+use crate::server::dask::key::{dask_key_ref_to_str, dask_key_ref_to_string, DaskKey, DaskKeyRef};
+use crate::server::gateway::Gateway;
+use crate::server::notifications::{ErrorNotification, Notifications};
+use crate::server::protocol::messages::worker::ToWorkerMessage;
+use crate::server::protocol::messages::worker::{
+    NewWorkerMsg, StealResponse, StealResponseMsg, TaskFinishedMsg,
 };
-use crate::server::protocol::messages::worker::{StealResponse, StealResponseMsg, TaskFinishedMsg};
 use crate::server::task::{DataInfo, ErrorInfo, Task, TaskRef, TaskRuntimeState};
 use crate::server::worker::WorkerRef;
 use crate::trace::{
-    trace_task_assign, trace_task_finish, trace_task_place, trace_task_remove, trace_worker_new,
-    trace_worker_steal, trace_worker_steal_response, trace_worker_steal_response_missing,
+    trace_task_assign, trace_task_finish, trace_task_new, trace_task_place, trace_task_remove,
+    trace_worker_new, trace_worker_steal, trace_worker_steal_response,
+    trace_worker_steal_response_missing,
 };
 
-impl Identifiable for Client {
-    type Id = ClientId;
-    type Key = DaskKey;
-
-    #[inline]
-    fn get_id(&self) -> Self::Id {
-        self.id()
-    }
-
-    #[inline]
-    fn get_key(&self) -> Self::Key {
-        self.key().into()
-    }
-}
-
-impl Identifiable for WorkerRef {
-    type Id = WorkerId;
-    type Key = DaskKey;
-
-    #[inline]
-    fn get_id(&self) -> Self::Id {
-        self.get().id()
-    }
-
-    #[inline]
-    fn get_key(&self) -> Self::Key {
-        self.get().key().into()
-    }
-}
-
 pub struct Core {
-    tasks_by_id: Map<TaskId, TaskRef>,
-    tasks_by_key: Map<DaskKey, TaskRef>,
-    clients: KeyIdMap<ClientId, Client, DaskKey>,
-    workers: KeyIdMap<WorkerId, WorkerRef, DaskKey>,
+    tasks: Map<TaskId, TaskRef>,
+    workers: Map<WorkerId, WorkerRef>,
 
     task_id_counter: IdCounter,
     worker_id_counter: IdCounter,
-    client_id_counter: IdCounter,
-    uid: DaskKey,
 
     scatter_counter: usize,
+
+    pub gateway: Box<dyn Gateway>,
```

Comment:
Genericý parametr byl moc otravný?

Reply:
Main reason is compiletime/size of code, bucase core is everywhere and in case of more gateways it would practically multiply the size of code.

Reply:
Moreoever I do not expect that gateway should be performance critical.